File: Python/examples/visualization/basketoption.py
# Example of option baskets
# Distributed under BSD License

import QuantLib as ql


class BasketOptionClass:
    def __init__(self, btype):
        # global data
        self.todaysDate = ql.Date(15, ql.May, 1998)
        ql.Settings.instance().evaluationDate = self.todaysDate
        self.settlementDate = ql.Date(17, ql.May, 1998)
        self.riskFreeQuote = ql.SimpleQuote(0.05)
        self.riskFreeRate = ql.FlatForward(self.settlementDate, ql.QuoteHandle(self.riskFreeQuote), ql.Actual365Fixed())

        # option parameters
        self.exercise = ql.EuropeanExercise(ql.Date(17, ql.May, 1999))
        self.payoff = ql.PlainVanillaPayoff(ql.Option.Call, 8.0)

        # market data
        self.underlying1 = ql.SimpleQuote(10.0)
        self.volatility1 = ql.BlackConstantVol(self.todaysDate, 0.20, ql.Actual365Fixed())
        self.dividendYield1 = ql.FlatForward(self.settlementDate, 0.05, ql.Actual365Fixed())
        self.underlying2 = ql.SimpleQuote(7.0)
        self.volatility2 = ql.BlackConstantVol(self.todaysDate, 0.10, ql.Actual365Fixed())
        self.dividendYield2 = ql.FlatForward(self.settlementDate, 0.05, ql.Actual365Fixed())

        self.process1 = ql.BlackScholesMertonProcess(
            ql.QuoteHandle(self.underlying1),
            ql.YieldTermStructureHandle(self.dividendYield1),
            ql.YieldTermStructureHandle(self.riskFreeRate),
            ql.BlackVolTermStructureHandle(self.volatility1),
        )

        self.process2 = ql.BlackScholesMertonProcess(
            ql.QuoteHandle(self.underlying2),
            ql.YieldTermStructureHandle(self.dividendYield2),
            ql.YieldTermStructureHandle(self.riskFreeRate),
            ql.BlackVolTermStructureHandle(self.volatility2),
        )

        self.procs = ql.StochasticProcessVector()
        self.procs.push_back(self.process1)
        self.procs.push_back(self.process2)

        self.matrix = ql.Matrix(2, 2)
        self.matrix[0][0] = 1.0
        self.matrix[1][1] = 1.0
        self.matrix[0][1] = 0.25
        self.matrix[1][0] = 0.25

        self.process = ql.StochasticProcessArray(self.procs, self.matrix)
        self.engine = ql.MCBasketEngine("lowdiscrepancy", timeStepsPerYear=1, requiredTolerance=0.02, seed=42)
        if btype == "min":
            bpayoff = ql.MinBasketPayoff(self.payoff)
        elif btype == "max":
            bpayoff = ql.MaxBasketPayoff(self.payoff)
        else:
            bpayoff = ql.AverageBasketPayoff(self.payoff, 2)

        self.option = ql.BasketOption(self.process, bpayoff, self.exercise, self.engine)

    def npv(self, x, y):
        self.underlying1.setValue(x)
        self.underlying2.setValue(y)
        return self.option.NPV()

    def setQuote(self, r):
        self.riskFreeQuote.setValue(r)

File: R/demo/basket-option.R

# inspired by python example with the same name

suppressMessages(library(QuantLib))

## global data
calendar <- TARGET()

todaysDate <- Date(15, "May", 1998)
invisible(Settings_instance()$setEvaluationDate(d=todaysDate))

settlementDays <- 3
settlementDate <- Calendar_advance(calendar, todaysDate, settlementDays, "Days")

cat('Today          : ', todaysDate$`__str__`(), "\n")
cat('Settlement Date: ', settlementDate$`__str__`(), "\n")

riskFreeRate = FlatForward(settlementDate, 0.05, Actual365Fixed())


# ### Option parameters
exerciseDate = Date(17, "May", 1999)
exercise = EuropeanExercise(exerciseDate)
payoff = PlainVanillaPayoff("Call", 8.0)


# ### Market data
underlying1 = SimpleQuote(7.0)
volatility1 = BlackConstantVol(todaysDate, calendar, 0.10, Actual365Fixed())
dividendYield1 = FlatForward(settlementDate, 0.05, Actual365Fixed())
underlying2 = SimpleQuote(7.0)
volatility2 = BlackConstantVol(todaysDate, calendar, 0.10, Actual365Fixed())
dividendYield2 = FlatForward(settlementDate, 0.05, Actual365Fixed())


process1 = BlackScholesMertonProcess(
    QuoteHandle(underlying1),
    YieldTermStructureHandle(dividendYield1),
    YieldTermStructureHandle(riskFreeRate),
    BlackVolTermStructureHandle(volatility1)
)

process2 = BlackScholesMertonProcess(
    QuoteHandle(underlying2),
    YieldTermStructureHandle(dividendYield2),
    YieldTermStructureHandle(riskFreeRate),
    BlackVolTermStructureHandle(volatility2)
)

corrMatrix = Matrix(2, 2)
invisible(Matrix_setitem(corrMatrix, 0, 0, 1.0))
invisible(Matrix_setitem(corrMatrix, 1, 1, 1.0))
invisible(Matrix_setitem(corrMatrix, 1, 0, 0.5))
invisible(Matrix_setitem(corrMatrix, 0, 1, 0.5))
cat(corrMatrix$`__str__`())

procVector = StochasticProcess1DVector(2)
invisible(StochasticProcess1DVector___setitem__(self = procVector, i = 0, x = process1))
invisible(StochasticProcess1DVector___setitem__(self = procVector, i = 1, x = process2))

process = StochasticProcessArray(array = procVector, correlation = corrMatrix)

# ### Pricing - European

basketoption = BasketOption(MaxBasketPayoff(payoff), exercise)
basketoption$setPricingEngine(
  MCPREuropeanBasketEngine(process=process, timeSteps=NA, timeStepsPerYear=1, brownianBridge=F, antitheticVariate=F, requiredSamples=NA, requiredTolerance=0.02, maxSamples=10000, seed=42)
)
print(basketoption$NPV())

basketoption = BasketOption(MinBasketPayoff(payoff), exercise)
basketoption$setPricingEngine(
  MCPREuropeanBasketEngine(process=process, timeSteps=NA, timeStepsPerYear=1, brownianBridge=F, antitheticVariate=F, requiredSamples=NA, requiredTolerance=0.02, maxSamples=10000, seed=42)
)
print(basketoption$NPV())

basketoption = BasketOption(AverageBasketPayoff(payoff, 2), exercise)
basketoption$setPricingEngine(
  MCPREuropeanBasketEngine(process=process, timeSteps=NA, timeStepsPerYear=1, brownianBridge=F, antitheticVariate=F, requiredSamples=NA, requiredTolerance=0.02, maxSamples=10000, seed=42)
)
print(basketoption$NPV())


# ### Pricing - American

americanExercise = AmericanExercise(settlementDate, exerciseDate)
americanbasketoption = BasketOption(MaxBasketPayoff(payoff), americanExercise)
americanbasketoption$setPricingEngine(
  MCPRAmericanBasketEngine(
        process=process,
        timeSteps=10,
        timeStepsPerYear=NA,
        brownianBridge=F,
        antitheticVariate=F,
        requiredSamples=50000,
        requiredTolerance=0.02,
        maxSamples=100000,
        seed=42,
        nCalibrationSamples=5000,
        polynomOrder=5,
        polynomType=LsmBasisSystem_Hermite_get()
    )
)
print(americanbasketoption$NPV())
